sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1134,7 +1138,7 @@ class Dataset[T] private[sql](
    */
   @scala.annotation.varargs
   def select(cols: Column*): DataFrame = withPlan {
-    Project(cols.map(_.named), logicalPlan)
+    Project(cols.map(_.named), AnalysisBarrier(logicalPlan))
```

Comment:
After this PR https://github.com/apache/spark/pull/11443, we always do the eager analysis.